resteasy-core/src/main/java/org/jboss/resteasy/specimpl/BuiltResponse.java

```diff
@@ -38,488 +31,254 @@
             {
-               if (!"".equals(str.trim()))
+               @Override
+               public byte[] getMarshalledBytes()
                {
-                  allowedMethods.add(str.trim().toUpperCase());
+                  return bytes;
                }
-            }
+
+               @Override
+               public Object getEntity()
+               {
+                  return finalObj;
+               }
+            };
          }
          else
          {
-            allowedMethods.add(toHeaderString(header).toUpperCase());
+            return finalObj;
          }
-      }
-
-      return allowedMethods;
-   }
-
-   protected String toHeaderString(Object header)
-   {
-      if (header instanceof String) return (String)header;
-      return getHeaderValueProcessor().toHeaderString(header);
-   }
 
-   @Override
-   public MultivaluedMap<String, String> getStringHeaders()
-   {
-      CaseInsensitiveMap<String> map = new CaseInsensitiveMap<String>();
-      for (Map.Entry<String, List<Object>> entry : metadata.entrySet())
+      }
+      catch (ProcessingException pe)
       {
-         for (Object obj : entry.getValue())
-         {
-            map.add(entry.getKey(), toHeaderString(obj));
-         }
+         throw pe;
       }
-      return map;
-   }
-
-   @Override
-   public String getHeaderString(String name)
-   {
-      List vals = metadata.get(name);
-      if (vals == null) return null;
-      StringBuilder builder = new StringBuilder();
-      boolean first = true;
-      for (Object val : vals)
+      catch (Exception ex)
       {
-         if (first) first = false;
-         else builder.append(",");
-         if (val == null) val = "";
-         val = toHeaderString(val);
-         if (val == null) val = "";
-         builder.append(val);
+         throw new ProcessingException(ex);
       }
-      return builder.toString();
    }
 
-   @Override
-   public URI getLocation()
+   protected InputStream getEntityStream()
    {
-      Object uri = metadata.getFirst(HttpHeaders.LOCATION);
-      if (uri == null) return null;
-      if (uri instanceof URI) return (URI)uri;
-      String str = null;
-      if (uri instanceof String) str = (String)uri;
-      else str = toHeaderString(uri);
-      return URI.create(str);
+      if (bufferedEntity != null) {
+         return new ByteArrayInputStream(bufferedEntity);
+      }
+      if (isClosed()) {
+         throw new ProcessingException(Messages.MESSAGES.streamIsClosed());
+      }
+      InputStream is = getInputStream();
+      return is != null ? new AbstractBuiltResponse.InputStreamWrapper<BuiltResponse>(is, this) : null;
    }
 
-   @Override
-   public Set<Link> getLinks()
+   protected void setInputStream(InputStream is)
    {
-      LinkHeaders linkHeaders = getLinkHeaders();
-      Set<Link> links = new HashSet<Link>();
-      links.addAll(linkHeaders.getLinks());
-      return links;
+      this.is = is;

```

Comment:
Hi @rsearls,

I see that setInputStream() in ApacheHttpClient43Engine is
```
         protected void setInputStream(InputStream is)
         {
            stream = is;
            resetEntity();
         }
```
Does it make sense to have resetEntity() here?

-Ron

Reply:
Like @asoldano said, there seem to be a lot of purely stylistic changes. For example,
```
@Override
public synchronized Object getEntity()
{
   abortIfClosed();
   Object entity = super.getEntity();
   if (entity != null)
   {
      return checkEntityReadAsInputStreamFullyConsumed(entity);
   }
   return checkEntityReadAsInputStreamFullyConsumed(getEntityStream());
}
```
starts in column 0. Not a big deal, but it increases the degree of difficulty ...

Reply:
The ClientResponse object in ApacheHttpClient43Engine is sometimes
called on a "get()" request which results in the input stream being
read and variable "streamFullyRead" set to TRUE.  This value needs to
be FALSE when a "getEntity()" call is make.  See testsuite case 
ClientFilterResponseBuilderTest and class ResponseBuilderCustomResponseFilter.
When I remove the call to resetEntity(); this test breaks.

This call could be replaced with setStreamFullyRead(false), but there
might be other use cases in which any of the other 4 variables in
resetEntity() need to be reset.

I would recommend leaving this as is.

Reply:
Sorry, I seem to be having trouble expressing myself clearly these days. I meant to say, "Does it make sense to add resetEntity() to BuiltResponse.setInputStream()?"

Reply:
This could be done.  None of our tests are negatively affected by adding it.   But do we really need it?  Customers have never reported any failures with the existing code and this code has been around a long time.  

Do you want me to add it any way?

Reply:
re: "Customers have never reported any failures with the existing code and this code has been around a long time."

Hmmm. Prior to this pull request, setInputStream() was abstract in ClientResponse, so it was implemented only in classes derived from ClientResponse. So its implementation in BuiltResponse is something new. Given that resetting the inputStream, in effect, changes the content of the response, doesn't it make sense to reset the entity? WDYT?